### ISSUE TITLE
Resolve woocommerce hook woocommerce_email_footer function causing recursion

### ISSUE THREAD

**Issue by shosaman** [woocommerce/woocommerce#24035]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

Here's my Issue:

I'm using Woocommerce hook woocommerce_email_footer() and inside my function I need to call $email->get_content() which causes recursion and PHP memory error and wordpress exits with system error

Tried to remove the hook before calling the $email->get_content() and adding the hook back right after this call. However this may not be fool proof solution since some other session which hits exactly at the time when my function has removed the action might totally miss the hook custom action

I've written the following code in functions.php of my theme to capture the mail contents (mail body) into a local file just before it is being sent whenever a new order is received
//
// Capture the contents of the Emails sent and save into local files 
// These Local files are used for further messaging through different channels 
//
function Save_Email_Contents_into_Local_File ( $email ) { 
   if ( $email->id == 'customer_processing_order' ) {

// Remove the action temporarily so as not to cause Recursion while we refer to $email functions
     remove_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 

    $TargetFilename = '/home/users/....../Sent_Mail.html' ;

      $html_message = $email->get_content();
      $formatted_message = $email->style_inline($html_message);
      file_put_contents($TargetFilename, $formatted_message);
 }

// Put the action back to original state
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 
}; 

// add the action 
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 

Please note in the above function, I'm referring the $email->get_content() public function. 

I'm using Woocommerce hook woocommerce_email_footer() and inside my function I need to call $email->get_content() which causes recursion and PHP memory error and wordpress exits with system error

Tried to remove the hook before calling the $email->get_content() and adding the hook back right after this call. However this may not be fool proof solution since some other session which hits exactly at the time when my function has removed the action might totally miss the hook custom action

I've written the following code in functions.php of my theme to capture the mail contents (mail body) into a local file just before it is being sent whenever a new order is received
// // Capture the contents of the Emails sent and save into local files // These Local files are used for further messaging through different channels //

function Save_Email_Contents_into_Local_File ( $email ) { 
   if ( $email->id == 'customer_processing_order' ) {

// Remove the action temporarily so as not to cause Recursion while we refer to $email functions
     remove_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 

    $TargetFilename = '/home/users/....../Sent_Mail.html' ;

      $html_message = $email->get_content();
      $formatted_message = $email->style_inline($html_message);
      file_put_contents($TargetFilename, $formatted_message);
 }

// Put the action back to original state
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 
}; 

// add the action 
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); 

Please note in the above function, I'm referring the $email->get_content() public function. 
If I do not do the remove_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 ); this function becomes recursive and fails with a PHP memory error.

Although this is a workable solution, Removing the action can potentially cause another instance of the 'customer_processing_order' from another user to miss the action and not come to this function if that session hits exactly at the time when this current session has removed the action and before adding the action again.

I'm sure I'm not doing it right! Is there any better way to accomplish what I need - basically I need the exact formatted mail content to be stored in a local file whenever the order is received. Similarly I will need local file stored for Order completion and order hold etc. but at later point of time.

Want to achieve storing formatted email into a local file a) without causing recursion / PHP memory errors b) without having to miss some instances of the execution missing the custom code attached to the hook.

## Expected/actual behavior

When I use woocommerce_email_footer hook which send me $email (type WC_Email) as the only parameter. If I use the $email->get_contents() function it causes recursion. 

Recursion is happening because when I make a call to $email->get_content() which is trying to get me the mail body including 'footers' once again encounters the hook due to the footer and comes inside my function again.. repeats. 

I was expecting to see no recursion.. 

## Isolating the problem

- [X ] This bug happens with only WooCommerce plugin active
- [X ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [X ] I can reproduce this bug consistently using the steps above

## WordPress Environment
5.2.2
<details>
```
Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin here.
```
</details>

**Comment by claudiosanches** [comment]
> I was expecting to see no recursion..

Sorry, but not possible in this conditions, we can't stop recursion on it, since `get_content()` needs to call footer, and footer needs to run that hook.
Sounds like you should try anything else instead of `$email->get_content()`, not clear in this thread why you need to call it, but probably you can find alternatives.

**Comment by shosaman** [comment]
Hi Caludio, Thank you for your gracious help.
I'm calling get_content() because I need the entire mail content to be captured and stored into a local file on the server just before WC send out the mail.
The only place where I felt it is most appropriate is when WC is done building the footer

In the spirit of our conversations around WC I want your small help..

I've found a workaround to solve the recursion, like the following. In my function I first UN-HOOK my function and finish my work and then RE-HOOK before I go out of my function

i.e.
//
// Capture the contents of the Emails sent and save into local files
// These Local files are used for further messaging through different channels
//

function Save_Email_Contents_into_Local_File ( $email ) {
if ( $email->id == 'customer_processing_order' ) {

// Remove the action temporarily so as not to cause Recursion while we refer to $email functions
remove_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 );
---
--
--- DO ALL PROCESSING INCLUDING CALLING $EMAIL->GET_CONTENT(); 
--- WHICH WORKS SINCE I'VE UN-HOOKED my FUNCTION
---  WITH THIS CODE THERE IS RECURSION...
---
}

// Put the action back to original state
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 );
};

// add the action
add_action( 'woocommerce_email_footer', 'Save_Email_Contents_into_Local_File', 20, 1 );

The workaround is: Within my function I'm UN-HOOKING the WC Footer action and before my function ends I'm RE-HOOKING the WC-Footer action

Do you see any issue with the above code? 

**My suspicion is that in a multi user environment when one of the user's thread has just un-hooked and inside my function - just at the same timeuser's thread hits the email footer - will it skip my function ?**

OR

In WC every user session will have its own session memory that is independent of the other user's session and even if the program at this time executing in one user's thread un-hooks other users' flow works seamlessly fine ?

**PLEASE help me understand how WC does session management across multiple users ? Answer to this will help analyze whether the above workaround is a proper solution or not.**

So far the above workaround solution has worked, and not causing recursion, but I'm not sure if there are several hundreds of users hit the system in parallel how does it behave - I do not know. Please help me understand how will this code work in such a situation.

Thank you much!
Sreenivasa Hosaamane

**Comment by claudiosanches** [comment]
@shosaman

Please let's try keep this small, becoming hard to follow.
So let's address it first.

> I'm calling get_content() because I need the entire mail content to be captured and stored into a local file on the server just before WC send out the mail.
> The only place where I felt it is most appropriate is when WC is done building the footer

This still sounds like a hack, and this issues is trying to allow this hack to works... Instead we should introduce a new hook, so no more need for hacks or incomplete solutions.
So where do you need a hook?

**Comment by shosaman** [comment]
Dear Claudio, Thank you much! Let me explain my need:
1. Once Order is successfully created, WC sends Order Received eMail to customer
2. I need to take that mail content and send that mail content also to the customer's Messenger (Telegram / whatsapp) as a Telegram message
**3. I'm saving the WC mail content into a local html file** 
4. Later I pull this html file through a python code and send the html file to Customer's Telegram. 

How can I achieve step 3 - please suggest any hook or available hooks or what new hook is possible that I can use ? 
Hope this is short and to the point..

**Comment by shosaman** [comment]
SO I need a hook just at the time when WC sends the mail - WC_EMAIL class passed as an argument  to my function

**Comment by shosaman** [comment]
SO Am I clear now ? 
I need a hook just at the time when WC sends the mail - WC_EMAIL class passed as an argument to my function
